### ISSUE TITLE
SSH-Backend Problems

### ISSUE THREAD

**Issue by INeido** [offen/docker-volume-backup#350]
**What are you trying to do?**
Use SSH as a backend.

**What is your current configuration?**
compose:
```
backup:
    container_name: co_backup
    image: offen/docker-volume-backup
    volumes:
      - /mnt/tank/docker/stacks/code:/etc/dockervolumebackup/conf.d
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
      - /mnt/tank/docker/stacks/code/code:/backup/co_code:ro
    restart: unless-stopped
```
.env:
```
# Code
BACKUP_SOURCES="/backup/co_code"
BACKUP_FILENAME="code.tar.gz"
BACKUP_CRON_EXPRESSION="*/1 * * * *"
BACKUP_RETENTION_DAYS="30"
SSH_HOST_NAME="192.168.1.3"
SSH_PORT="13378"
SSH_USER="backup"
SSH_PASSWORD="CENSORED"
SSH_REMOTE_PATH="/volume1/backup/code"
```

**Log output**
```
time=2024-02-12T11:45:21.219-08:00 level=INFO msg="Successfully scheduled backup backup.env with expression */1 * * * *"

time=2024-02-12T11:46:00.036-08:00 level=INFO msg="Now running script on schedule */1 * * * *"

time=2024-02-12T11:46:00.166-08:00 level=ERROR msg="Unexpected error running schedule */1 * * * *: runScript: error instantiating script: newScript: error creating ssh storage backend: NewStorageBackend: error creating sftp client: ssh: subsystem request failed" error="runScript: error instantiating script: newScript: error creating ssh storage backend: NewStorageBackend: error creating sftp client: ssh: subsystem request failed"
```

**Additional context**
I can access the server through SSH using a terminal with the same login as provided above. The SSH-Server is a Synology NAS, if that matters at all.
Am I missing something, is this not how SSH was intended?

**Comment by m90** [comment]
I think you will still need to mount the key you want to use to authenticate.

See: https://offen.github.io/docker-volume-backup/recipes/#backing-up-to-ssh

and the test case for the `ssh` storage: https://github.com/offen/docker-volume-backup/blob/4c74313222e3a0f06b7aa87b672c703735babdb7/test/ssh/docker-compose.yml

**Comment by INeido** [comment]
I set up a keypair and got it working in principle (so using a terminal on another pc), but still get the same error:

```
time=2024-02-12T14:08:00.418-08:00 level=ERROR msg="Unexpected error running schedule */1 * * * *: runScript: error instantiating script: newScript: error creating ssh storage backend: NewStorageBackend: error creating sftp client: ssh: subsystem request failed" error="runScript: error instantiating script: newScript: error creating ssh storage backend: NewStorageBackend: error creating sftp client: ssh: subsystem request failed"
```

New config looks like this:
```
# Code
BACKUP_SOURCES="/backup/co_code"
BACKUP_FILENAME="code.tar.gz"
BACKUP_CRON_EXPRESSION="*/1 * * * *"
BACKUP_RETENTION_DAYS="30"
SSH_HOST_NAME="192.168.1.3"
SSH_PORT="13378"
SSH_USER="backup"
SSH_IDENTITY_PASSPHRASE="CENSORED"
SSH_REMOTE_PATH="/volume1/backup/code"
```

```
backup:
    container_name: co_backup
    image: offen/docker-volume-backup
    volumes:
      - /mnt/tank/docker/stacks/code:/etc/dockervolumebackup/conf.d
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
      - /mnt/tank/docker/backup/id_rsa:/root/.ssh/id_rsa
      - /mnt/tank/docker/stacks/code/code:/backup/co_code:ro
    restart: unless-stopped
```

**Comment by INeido** [comment]
Found the problem! I was using the Port for SSH, but on my Host I have a different port for SFTP. I switched the port and now I don't get that error anymore...

...rather a different one:

```
time=2024-02-12T14:20:28.453-08:00 level=INFO msg="/etc/dockervolumebackup/conf.d was found, using configuration files from this directory."

time=2024-02-12T14:20:28.453-08:00 level=INFO msg="Successfully scheduled backup backup.env with expression */1 * * * *"

time=2024-02-12T14:21:00.031-08:00 level=INFO msg="Now running script on schedule */1 * * * *"

time=2024-02-12T14:21:00.309-08:00 level=INFO msg="Created backup of `/backup/co_code` at `/tmp/code.tar.gz`."

time=2024-02-12T14:21:00.320-08:00 level=INFO msg="Removed tar file `/tmp/code.tar.gz`."

time=2024-02-12T14:21:00.320-08:00 level=ERROR msg="Script run failed: runScript: error running script: copyArchive: error copying archive: (*sshStorage).Copy: error creating file: file does not exist" error="runScript: error running script: copyArchive: error copying archive: (*sshStorage).Copy: error creating file: file does not exist"

time=2024-02-12T14:21:00.320-08:00 level=ERROR msg="Unexpected error running schedule */1 * * * *: runScript: error running script: copyArchive: error copying archive: (*sshStorage).Copy: error creating file: file does not exist" error="runScript: error running script: copyArchive: error copying archive: (*sshStorage).Copy: error creating file: file does not exist"
```

I do have to say, it does look weird. Is the tar first removed and only afterwards does it try to copy it?

**Comment by m90** [comment]
The log ordering is a bit confusing, yes, but is correct: before logging the error that made the program halt, cleanup routines (e.g. deleting files) will run and might also log things. This might be very misleading in the context of your setup, but it's not the cause of the error.

Your error points here: https://github.com/offen/docker-volume-backup/blob/aab47509d988ee042402847e392112061b636e6c/internal/storage/ssh/ssh.go#L118-L121 which makes me think the path you specify in `SSH_REMOTE_PATH` does not exist.

**Comment by m90** [comment]
> I have to set the SSH parameters in the .env file, otherwise it will not use the backend. Is that intended or am I doing something wrong?

This is a bug that was just introduced in `v2.37.0`, which was released yesterday (#357). For now, the solution for you should probably be downgrading until this isi fixed (you might need to adjust your cron schedule though, not sure).

I'm closing this issue as your configuration problems seem to have been resolved.

**Comment by m90** [comment]
The bug you've been encountering is now fixed, so you should be able to use the latest version (v2.37.2) again.
